packages/java/endpoint/src/test/java/dev/hilla/EndpointInvokerTest.java

```diff
@@ -0,0 +1,172 @@
+            protected EndpointAccessChecker getAccessChecker() {
+                return endpointAccessChecker;
+            }
+        };
+    }
+
+    @Test
+    public void when_invokedMethod_isDeclaredIn_Endpoint_accessCheckingIsDoneBasedOn_EndpointMethod()
+            throws Exception {
+
+        @EndpointExposed
+        class ParentEndpoint {
+            public String sayHello() {
+                return "Hello from ParentEndpoint";
+            }
+        }
+
+        @Endpoint
+        class TestEndpoint extends ParentEndpoint {
+            public void test() {
+            }
+        }
+
+        TestEndpoint test = new TestEndpoint();
+
+        endpointRegistry.registerEndpoint(test);
+
+        endpointInvoker.invoke("TestEndpoint", "test", body, principal,
+                requestMock::isUserInRole);
+
+        Mockito.verify(endpointAccessChecker, Mockito.times(1))
+                .check(any(Method.class), any(), any());
+        Mockito.verify(endpointAccessChecker, Mockito.times(0))
+                .check(any(Class.class), any(), any());
+    }
+
+    @Test
+    public void when_invokedMethod_isDeclaredIn_EndpointExposed_accessCheckingIsDoneBasedOn_EndpointClass()
+            throws Exception {
+
+        @EndpointExposed
+        class ParentEndpoint {
+            public String sayHello() {
+                return "Hello from ParentEndpoint";
+            }
+        }
+
+        @Endpoint
+        class TestEndpoint extends ParentEndpoint {
+            public void test() {
+            }
+        }
+
+        TestEndpoint test = new TestEndpoint();
+
+        endpointRegistry.registerEndpoint(test);
+
+        endpointInvoker.invoke("TestEndpoint", "sayhello", body, principal,
+                requestMock::isUserInRole);
+
+        Mockito.verify(endpointAccessChecker, Mockito.times(0))
+                .check(any(Method.class), any(), any());
+        Mockito.verify(endpointAccessChecker, Mockito.times(1))
+                .check(any(Class.class), any(), any());
+    }
+
+    @Test
+    public void when_invokedMethod_isOverriddenIn_Endpoint_accessCheckingIsDoneBasedOn_EndpointMethod()
+            throws Exception {
+
+        @EndpointExposed
+        class ParentEndpoint {
+            public String sayHello() {
+                return "Hello from ParentEndpoint";
+            }
+        }
+
+        @Endpoint
+        class TestEndpoint extends ParentEndpoint {
+
+            @Override
+            public String sayHello() {
+                return "Hello from TestEndpoint";
+            }
+
+            public void test() {
+            }
+        }
+
+        TestEndpoint test = new TestEndpoint();
+
+        endpointRegistry.registerEndpoint(test);
+
+        endpointInvoker.invoke("TestEndpoint", "sayhello", body, principal,
+                requestMock::isUserInRole);
+
+        Mockito.verify(endpointAccessChecker, Mockito.times(1))
+                .check(any(Method.class), any(), any());
```

Comment:
Would be nice to verify that the method being checked is `TestEndpoint::sayHello` and not `ParentEndpoint::sayHello`.